src/stan/services/pathfinder/single.hpp

```diff
@@ -0,0 +1,1104 @@
+ * @tparam Logger Type of logger callback
+ * @param lp_fun Functor to calculate the log density
+ * @param constrain_fun A functor to transform parameters to the constrained
+ * space
+ * @param[in,out] rng A generator to produce standard gaussian random variables
+ * @param taylor_approx The taylor approximation at this iteration of LBFGS
+ * @param num_samples Number of approximate samples to generate
+ * @param alpha The approximation of the diagonal hessian
+ * @param iter_msg The beginning of messages that includes the iteration number
+ * @param logger A callback writer for messages
+ */
+template <bool ReturnElbo = true, typename LPF, typename ConstrainF,
+          typename RNG, typename EigVec, typename Logger>
+inline elbo_est_t est_approx_draws(LPF&& lp_fun, ConstrainF&& constrain_fun,
+                                   RNG&& rng,
+                                   const taylor_approx_t& taylor_approx,
+                                   size_t num_samples, const EigVec& alpha,
+                                   const std::string& iter_msg,
+                                   Logger&& logger) {
+  boost::variate_generator<boost::ecuyer1988&, boost::normal_distribution<>>
+      rand_unit_gaus(rng, boost::normal_distribution<>());
+  const auto num_params = taylor_approx.x_center.size();
+  size_t lp_fun_calls = 0;
+  Eigen::MatrixXd unit_samps
+      = generate_matrix(rand_unit_gaus, num_params, num_samples);
+  Eigen::Array<double, -1, -1> lp_mat(num_samples, 2);
+  lp_mat.col(0) = (-taylor_approx.logdetcholHk)
+                  + -0.5
+                        * (unit_samps.array().square().colwise().sum()
+                           + num_params * stan::math::LOG_TWO_PI);
+  Eigen::MatrixXd approx_samples
+      = approximate_samples(std::move(unit_samps), taylor_approx);
+  debug::rnorm_draws(logger, approx_samples);
+  Eigen::VectorXd approx_samples_col;
+  std::stringstream pathfinder_ss;
+  const auto log_stream = [](auto& logger, auto& pathfinder_ss) {
+    if (pathfinder_ss.str().length() > 0) {
+      logger.info(pathfinder_ss);
+      pathfinder_ss.str(std::string());
+    }
+  };
+  for (Eigen::Index i = 0; i < num_samples; ++i) {
+    try {
+      approx_samples_col = approx_samples.col(i);
+      ++lp_fun_calls;
+      lp_mat.coeffRef(i, 1) = lp_fun(approx_samples_col, pathfinder_ss);
+      log_stream(logger, pathfinder_ss);
+    } catch (const std::exception& e) {
+      lp_mat.coeffRef(i, 1) = -std::numeric_limits<double>::infinity();
+      log_stream(logger, pathfinder_ss);
+    }
+  }
+  Eigen::Array<double, -1, 1> lp_ratio = (lp_mat.col(1)) - lp_mat.col(0);
+  if (ReturnElbo) {
+    double ELBO = lp_ratio.mean();
+    debug::elbo_draws(logger, taylor_approx, approx_samples, lp_mat, ELBO);
+    return elbo_est_t{ELBO, lp_fun_calls, std::move(approx_samples),
+                      std::move(lp_mat), std::move(lp_ratio)};
+  } else {
+    return elbo_est_t{-std::numeric_limits<double>::infinity(), lp_fun_calls,
```

Comment:
The if statement here is a small optimization thing. `ReturnElbo` is a compile time constant so dead code elimination removes the unused branch. A ternary operator, even with a compile time constant, would cause both sides to be evaluated.